### PR TITLE
Adding goals support to the PHP SDK

### PR DESCRIPTION
**Pull request by nickveenhof**

### CODE REVIEW

**Comment by nickveenhof** [comment]
@ynx , review please? :)

File: src/Entity/Goal.php [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+namespace Acquia\LiftClient\Entity;
+
+use Acquia\LiftClient\Exception\LiftSdkException;
+use Acquia\LiftClient\Utility\Utility;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Normalizer\CustomNormalizer;
+
+class Goal extends \ArrayObject
```
Comment (ynx) [inline src/Entity/Goal.php:11]: Right now, class Goal/Slot/Visibility extends \ArrayObject, and uses EntityValueTrait/EntityTrait to share common functions. Is it possible that we define an abstract Acquia\LiftClient\Entity\Entity that extends \ArrayObject, and then having it encapsulate all the shared functions and logic? In this way, class Goal/Slot/Visibility can all just extend Entity.

File: src/Entity/Goal.php [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        $this['name'] = $name;
+
+        return $this;
+    }
+
+    /**
+     * Gets the 'name' parameter.
+     *
+     * @return string
+     *                The Name of the Goal
```
Comment (ynx) [inline src/Entity/Goal.php]: Misaligned?

File: src/Entity/Goal.php [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+     *
+     * @return \Acquia\LiftClient\Entity\Goal
+     */
+    public function setSiteIds(array $siteIds)
+    {
+        if (Utility::arrayDepth($siteIds) > 1) {
+            throw new LiftSdkException('site_ids argument is more than 1 level deep.');
+        }
+
+        // Set only the array values to the rule_ids property.
```
Comment (ynx) [inline src/Entity/Goal.php]: Should probably be "site_ids".

File: src/Entity/Goal.php [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+     * @return \Acquia\LiftClient\Entity\Goal
+     */
+    public function setRuleIds(array $ruleIds)
+    {
+        if (Utility::arrayDepth($ruleIds) > 1) {
+            throw new LiftSdkException('Rule Ids argument is more than 1 level deep.');
+        }
+
+        // Set only the array values to the rule_ids property.
+        $this['rule_ids'] = array_values($ruleIds);
```
Comment (ynx) [inline src/Entity/Goal.php:126]: You probably already know all these - Drupal 8 uses snake_case variable names and camelCased function names, which is following [PHP official recommendation](http://git.php.net/?p=php-src.git;a=blob_plain;f=CODING_STANDARDS;hb=HEAD). However, the vast majority PHP community has moved towards all camelCase (Symfony/Zend/Laravel and pretty much all of the ones that matter). Given the fact that this is a new, flashy, general-purpose, and dev-facing SDK, I'd personally just use camelCase for all. So I'd recommend no "rule_ids", just "ruleIds".

File: src/Manager/GoalManager.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     *
+     * @param array $id
+     *
+     * @throws \GuzzleHttp\Exception\RequestException
+     *
+     * @return \Acquia\LiftClient\Entity\Goal
+     */
+    public function get(
+      $id
+    ) {
```
Comment (ynx) [inline src/Manager/GoalManager.php]: Misaligned? And more of these below.

File: src/Entity/Goal.php [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    /**
+     * Gets the 'rule_ids' parameter.
+     *
+     * @return array The Rule Identifiers
+     */
+    public function getRuleIds()
+    {
+        return $this->getEntityValue('rule_ids', array());
```
Comment (nickveenhof) [inline src/Entity/Goal.php:138]: Function description mentioned that second param has to be a string, this is solved in the rule endpoint branch

File: src/Utility/Utility.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+              if ($depth > $max_depth) {
+                  $max_depth = $depth;
+              }
+          }
+      }
+
+      return $max_depth;
+  }
+}
```
Comment (ynx) [inline src/Utility/Utility.php:33]: Could use a unit test since logic looks like it has complexity.

File: test/GoalTest.php [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        $this->assertEquals($response->getRuleIds(), $goal->getRuleIds());
+
+        // Check if the site_ids is equal.
+        $this->assertEquals($response->getSiteIds(), $goal->getSiteIds());
+
+        // Check if the site_ids is equal.
+        $this->assertEquals($response->getEventNames(), $goal->getEventNames());
+
+        // Check if the global was set correctly.
+        $this->assertEquals($response->getGlobal(), $goal->getGlobal());
```
Comment (ynx) [inline test/GoalTest.php]: I was taught once that it is better to use a plain value (like a plain string) in the "expected" field. No biggie either way though.

**Comment by ynx** [comment]
Awesomely organized code. I left some minor comments but no blocker.

File: src/Entity/Goal.php (hunk 4 above)
Comment (nickveenhof) [reply src/Entity/Goal.php:126]: rule_ids is the API spec, not a PHP thing :)

File: src/Manager/GoalManager.php (hunk 5 above)
Comment (nickveenhof) [reply src/Manager/GoalManager.php]: The syntax checker did this, I swear! Curious whether this is some official formatting. I'll revert

File: src/Entity/Goal.php (hunk 1 above)
Comment (nickveenhof) [reply src/Entity/Goal.php:11]: Hmm, I could do that indeed. I'll move the Trait getEntityValue to that class and remove the Trait. I won't do it in this PR, as I have a couple of other PRs still to go but I'll incorporate it there

File: src/Manager/GoalManager.php (hunk 5 above)
Comment (ynx) [reply src/Manager/GoalManager.php]: Cool!

File: src/Entity/Goal.php (hunk 1 above)
Comment (ynx) [reply src/Entity/Goal.php:11]: Sound great!

File: src/Entity/Goal.php (hunk 2 above)
Comment (nickveenhof) [reply src/Entity/Goal.php]: Fixed

File: src/Entity/Goal.php (hunk 3 above)
Comment (nickveenhof) [reply src/Entity/Goal.php]: Fixed

File: src/Manager/GoalManager.php (hunk 5 above)
Comment (nickveenhof) [reply src/Manager/GoalManager.php]: Fixed

File: src/Utility/Utility.php (hunk 7 above)
Comment (nickveenhof) [reply src/Utility/Utility.php:33]: Fixed

File: test/GoalTest.php (hunk 8 above)
Comment (nickveenhof) [reply test/GoalTest.php]: Hmm, that makes sense. I'll change it

**Pull request merged**